File: MarkdownEditor/MarkdownEditorTests/MarkdownRendererConverterTests.m
//
//  MarkdownRendererConverterTests.m
//  MarkdownEditorTests
//

#import <XCTest/XCTest.h>
#import "../MarkdownEditor/Sources/Converter/MarkdownRendererConverter.h"

@interface MarkdownRendererConverterTests : XCTestCase
@end

@implementation MarkdownRendererConverterTests

- (void)testExposesFormatAndCss {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    XCTAssertEqualObjects(converter.title, @"Title");
    XCTAssertEqualObjects(converter.format, @"gfm");
    XCTAssertEqualObjects(converter.css, @"style.css");
    XCTAssertNil(converter.header);
}

- (void)testRendersMarkdownContainingApostrophesWithoutShellEscaping {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"# Bob's Notes"];

    XCTAssertTrue([converter.html containsString:@"Bob"]);
    XCTAssertTrue([converter.html containsString:@"Notes"]);
    XCTAssertFalse([converter.html containsString:@"unexpected EOF"]);
}

- (void)testRawHTMLIsNotRenderedAsExecutableMarkup {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"<script>alert('x')</script>\n\n<img src=\"https://example.com/tracker.png\">"];

    XCTAssertFalse([converter.html containsString:@"<script>"]);
    XCTAssertFalse([converter.html containsString:@"<img src=\"https://example.com/tracker.png\">"]);
    XCTAssertTrue([converter.html containsString:@"&lt;script&gt;"] ||
                  [converter.html containsString:@"alert"]);
}

- (void)testDangerousMarkdownLinksAreRenderedWithoutHref {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"[Run JavaScript](javascript:alert(1))"];

    XCTAssertFalse([converter.html containsString:@"href=\"javascript:"]);
    XCTAssertTrue([converter.html containsString:@"Run JavaScript"]);
}

- (void)testRemoteMarkdownImagesAreNotRenderedAsImages {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"![Tracker](https://example.com/tracker.png)"];

    XCTAssertFalse([converter.html containsString:@"<img"]);
    XCTAssertFalse([converter.html containsString:@"https://example.com/tracker.png"]);
    XCTAssertTrue([converter.html containsString:@"Tracker"]);
}

- (void)testAbsoluteFileMarkdownImagesAreNotRenderedAsImages {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"![Secret](file:///Users/random/Documents/secret.png)"];

    XCTAssertFalse([converter.html containsString:@"<img"]);
    XCTAssertFalse([converter.html containsString:@"file:///Users/random/Documents/secret.png"]);
    XCTAssertTrue([converter.html containsString:@"Secret"]);
}

- (void)testRelativeMarkdownImagesAreRenderedAsImages {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];

    [converter setContentWithString:@"![Diagram](diagram.png)"];

    XCTAssertTrue([converter.html containsString:@"<img src=\"diagram.png\""]);
    XCTAssertTrue([converter.html containsString:@"alt=\"Diagram\""]);
}

- (void)testInlineFormattingMatchesMarkdownSyntax {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];
    NSDictionary<NSNumber *, NSString *> *expectations = @{
        @(TextConverterFormatBold): @"**text**",
        @(TextConverterFormatItalic): @"*text*",
        @(TextConverterFormatStrikeThrough): @"~~text~~",
        @(TextConverterFormatLink): @"[text](url)",
    };

    for (NSNumber *format in expectations) {
        XCTAssertEqualObjects([converter formattedStringWithString:@"text" format:format.unsignedIntegerValue],
                              expectations[format]);
    }
}

- (void)testBlockFormattingMatchesMarkdownSyntax {
    MarkdownRendererConverter *converter = [[MarkdownRendererConverter alloc] initWithTitle:@"Title"
                                                                 format:@"gfm"
                                                                 header:nil
                                                                    css:@"style.css"];
    NSDictionary<NSNumber *, NSString *> *expectations = @{
        @(TextConverterFormatCode): @"```\none\ntwo\n```",
        @(TextConverterFormatQuote): @"> one\n> two\n",
        @(TextConverterFormatListBulleted): @"- one\n- two\n",
        @(TextConverterFormatListNumbered): @"1. one\n1. two\n",
    };

    for (NSNumber *format in expectations) {
        XCTAssertEqualObjects([converter formattedStringWithString:@"one\ntwo" format:format.unsignedIntegerValue],
                              expectations[format]);
    }
}

@end

File: DraftmarkTests/MarkdownRendererConverterTests.swift
import XCTest
@testable import Draftmark

final class MarkdownRendererConverterTests: XCTestCase {
    func testExposesFormatAndCss() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        XCTAssertEqual(converter.title, "Title")
        XCTAssertEqual(converter.format, "gfm")
        XCTAssertEqual(converter.css, "style.css")
        XCTAssertNil(converter.header)
    }

    func testRendersMarkdownContainingApostrophesWithoutShellEscaping() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "# Bob's Notes")

        XCTAssertTrue(converter.html.contains("Bob"))
        XCTAssertTrue(converter.html.contains("Notes"))
        XCTAssertFalse(converter.html.contains("unexpected EOF"))
    }

    func testRawHTMLIsNotRenderedAsExecutableMarkup() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "<script>alert('x')</script>\n\n<img src=\"https://example.com/tracker.png\">")

        XCTAssertFalse(converter.html.contains("<script>"))
        XCTAssertFalse(converter.html.contains("<img src=\"https://example.com/tracker.png\">"))
        XCTAssertTrue(converter.html.contains("&lt;script&gt;") || converter.html.contains("alert"))
    }

    func testDangerousMarkdownLinksAreRenderedWithoutHref() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "[Run JavaScript](javascript:alert(1))")

        XCTAssertFalse(converter.html.contains("href=\"javascript:"))
        XCTAssertTrue(converter.html.contains("Run JavaScript"))
    }

    func testRemoteMarkdownImagesAreNotRenderedAsImages() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "![Tracker](https://example.com/tracker.png)")

        XCTAssertFalse(converter.html.contains("<img"))
        XCTAssertFalse(converter.html.contains("https://example.com/tracker.png"))
        XCTAssertTrue(converter.html.contains("Tracker"))
    }

    func testAbsoluteFileMarkdownImagesAreNotRenderedAsImages() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "![Secret](file:///Users/random/Documents/secret.png)")

        XCTAssertFalse(converter.html.contains("<img"))
        XCTAssertFalse(converter.html.contains("file:///Users/random/Documents/secret.png"))
        XCTAssertTrue(converter.html.contains("Secret"))
    }

    func testRelativeMarkdownImagesAreRenderedAsImages() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "![Diagram](diagram.png \"Architecture\")")

        XCTAssertTrue(converter.html.contains("<img src=\"diagram.png\""))
        XCTAssertTrue(converter.html.contains("alt=\"Diagram\""))
        XCTAssertTrue(converter.html.contains("title=\"Architecture\""))
    }

    func testInlineCodeLineBreaksAndInlineHTMLAreRenderedSafely() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")

        converter.setContent(with: "`let value = 1`\n\nfirst\\\nsecond\n\n<span>raw</span>")

        XCTAssertTrue(converter.html.contains("<code>let value = 1</code>"))
        XCTAssertTrue(converter.html.contains("first<br />"))
        XCTAssertTrue(converter.html.contains("&lt;span&gt;raw&lt;/span&gt;"))
    }

    func testSymbolLinksRenderAsEscapedCode() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: nil)

        converter.setContent(with: "See ``MarkdownHTMLRenderer``.")

        XCTAssertTrue(converter.html.contains("<code>MarkdownHTMLRenderer</code>"))
    }

    func testInlineFormattingMatchesMarkdownSyntax() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")
        let expectations: [TextConverterFormat: String] = [
            .bold: "**text**",
            .italic: "*text*",
            .strikeThrough: "~~text~~",
            .link: "[text](url)"
        ]

        for (format, expected) in expectations {
            XCTAssertEqual(converter.formattedString(with: "text", format: format), expected)
        }
    }

    func testBlockFormattingMatchesMarkdownSyntax() {
        let converter = MarkdownRendererConverter(title: "Title", format: "gfm", header: nil, css: "style.css")
        let expectations: [TextConverterFormat: String] = [
            .code: "```\none\ntwo\n```",
            .quote: "> one\n> two\n",
            .listBulleted: "- one\n- two\n",
            .listNumbered: "1. one\n1. two\n"
        ]

        for (format, expected) in expectations {
            XCTAssertEqual(converter.formattedString(with: "one\ntwo", format: format), expected)
        }
    }
}
